File: python/camera_ui_sdk/sensor/object.py
from __future__ import annotations

from abc import abstractmethod
from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType
from .detection import Detection, DetectionLabel, VideoFrameData
from .spec import ObjectModelSpec


class ObjectProperty(StrEnum):
    """Property names of an object detection sensor."""

    Detected = "detected"  # Whether any object is currently detected
    Detections = "detections"  # List of detected objects with labels and bounding boxes
    Labels = "labels"  # Unique labels of the current detections (auto-derived when reporting detections)


class TrackVelocity(TypedDict):
    """Signed centroid velocity vector in normalized units per frame step.

    Positive x = moving right, positive y = moving down. Consumers doing
    motion prediction (PTZ autotrack, trajectory estimation) should use this
    instead of deriving velocity from frame-to-frame position deltas.
    """

    x: float
    y: float


class TrackedDetection(Detection, total=False):
    """Detection enriched with tracking metadata (stable IDs, velocity)."""

    trackId: int  # Stable sequential ID for this object across frames
    trackAge: int  # Number of frames this object has been continuously tracked
    trackSpeed: float  # Velocity magnitude in normalized units per frame; 0 = stationary
    trackVelocity: TrackVelocity  # Signed centroid velocity vector in normalized units per frame
    trackLost: bool  # True if the object was not matched in the current frame


class ObjectSensorProperties(TypedDict):
    """Property shape carried by an ObjectSensor."""

    detected: bool
    detections: list[TrackedDetection]
    labels: list[DetectionLabel]


class ObjectPropertyChangeData(TypedDict):
    """Property change payload emitted on ObjectSensorLike.onPropertyChanged."""

    property: str  # ObjectProperty value
    value: bool | list[Detection] | list[DetectionLabel]


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class ObjectSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for an object detection sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Object

    @overload
    def getValue(self, property: Literal[ObjectProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: Literal[ObjectProperty.Detections]) -> list[TrackedDetection] | None: ...
    @overload
    def getValue(self, property: Literal[ObjectProperty.Labels]) -> list[DetectionLabel] | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...

    @property
    def onPropertyChanged(self) -> Observable[ObjectPropertyChangeData]: ...


class ObjectSensor(Sensor[ObjectSensorProperties, TStorage, str], Generic[TStorage]):
    """Object detection sensor that reports detected objects (person, vehicle, animal, etc.).

    Plugin authors call `reportDetections(list)` to push detection results.
    `detected` and `labels` are auto-derived from the detection list.
    """

    _requires_frames = False

    def __init__(self, name: str = "Object Sensor") -> None:
        super().__init__(name)
        self._write_state(
            {
                ObjectProperty.Detected.value: False,
                ObjectProperty.Detections.value: [],
                ObjectProperty.Labels.value: [],
            }
        )

    @property
    def type(self) -> SensorType:
        return SensorType.Object

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        """Whether any object is currently detected."""
        return bool(self.props.detected)

    @property
    def detections(self) -> list[TrackedDetection]:
        """Current detection list."""
        return self.props.detections or []

    @property
    def labels(self) -> list[DetectionLabel]:
        """Unique labels of the current detections."""
        return self.props.labels or []

    def reportDetections(
        self,
        detected: bool,
        detections: list[TrackedDetection] | None = None,
    ) -> None:
        """Report detected objects. Auto-derives ``detected`` and ``labels``
        from the list.

        - ``reportDetections(True)`` — something detected without specific data.
          The SDK synthesizes a single full-frame ``'motion'`` detection as a
          generic fallback.
        - ``reportDetections(True, [...])`` — explicit detections (typical case).
        - ``reportDetections(False)`` — clear.

        Args:
            detected: Whether any object is currently detected.
            detections: Optional explicit object detections (with optional
                tracking metadata).

        Example:
            ```python
            sensor.reportDetections(
                True,
                [
                    TrackedDetection(
                        label="person",
                        confidence=0.92,
                        box=BoundingBox(x=0.1, y=0.2, width=0.3, height=0.4),
                    )
                ],
            )
            sensor.reportDetections(False)
            ```
        """
        list_ = self._normalize_reported_detections(
            detected,
            detections,  # type: ignore[arg-type]
            "motion",
        )
        labels = list(dict.fromkeys(d["label"] for d in list_))
        self._write_state(
            {
                ObjectProperty.Detected.value: detected,
                ObjectProperty.Detections.value: list_,
                ObjectProperty.Labels.value: labels,
            }
        )

    def clearDetections(self) -> None:
        """Explicitly clear detection state (detected = False, detections = [], labels = [])."""
        self.reportDetections(False)

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored. State is reported via `reportDetections`."""
        # No-op — object detection state is reported by the plugin, not set externally.


class ObjectResult(TypedDict):
    """Return type for ObjectDetectorSensor.detectObjects()."""

    detected: bool  # Whether any object is detected in this frame
    detections: list[TrackedDetection]  # Detections emitted for this frame


class ObjectDetectorSensor(ObjectSensor[TStorage], Generic[TStorage]):
    """Object detector that receives video frames from the backend pipeline.

    Extend this class and implement ``detectObjects`` and ``modelSpec``. The
    backend scales frames to match ``modelSpec.input`` dimensions before each
    call.
    """

    _requires_frames = True

    @property
    @abstractmethod
    def modelSpec(self) -> ObjectModelSpec:
        """Declares the expected input dimensions. The backend scales frames to match."""
        ...

    @abstractmethod
    async def detectObjects(self, frame: VideoFrameData) -> ObjectResult:
        """Analyze a single video frame for objects. Called by the backend at the configured interval."""
        ...

File: node/src/sensor/object.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';
import type { Detection, DetectionLabel, VideoFrameData } from './detection.js';
import type { ObjectModelSpec } from './spec.js';

/**
 * Property names of an object detection sensor.
 *
 * @internal
 */
export enum ObjectProperty {
  /** Whether any object is currently detected. */
  Detected = 'detected',
  /** List of detected objects with labels and bounding boxes. */
  Detections = 'detections',
  /** Unique labels of the current detections (auto-derived when reporting detections). */
  Labels = 'labels',
}

/** Detection enriched with tracking metadata (stable IDs, velocity). */
export interface TrackedDetection extends Detection {
  /** Stable sequential ID for this object across frames. */
  trackId?: number;
  /** Number of frames this object has been continuously tracked. */
  trackAge?: number;
  /** Velocity magnitude in normalized units per frame. 0 = stationary. */
  trackSpeed?: number;
  /**
   * Signed centroid velocity in normalized units per frame. Positive x = moving right,
   * positive y = moving down. Prefer this over deriving velocity from position deltas.
   */
  trackVelocity?: { x: number; y: number };
  /** True if the object was not matched in the current frame. */
  trackLost?: boolean;
}

/**
 * Property values of an object detection sensor.
 *
 * @internal
 */
export interface ObjectSensorProperties {
  [ObjectProperty.Detected]: boolean;
  [ObjectProperty.Detections]: TrackedDetection[];
  [ObjectProperty.Labels]: DetectionLabel[];
}

/** Read-only proxy interface for an object detection sensor. */
export interface ObjectSensorLike extends SensorLike {
  readonly type: SensorType.Object;
  readonly onPropertyChanged: Observable<PropertyChangeOf<ObjectSensorProperties>>;

  getValue(property: ObjectProperty.Detected): boolean | undefined;
  getValue(property: ObjectProperty.Detections): TrackedDetection[] | undefined;
  getValue(property: ObjectProperty.Labels): DetectionLabel[] | undefined;
  getValue(property: string): unknown;
}

/**
 * Object detection sensor that reports detected objects (person, vehicle, animal, etc.).
 *
 * Plugin authors call `reportDetections(list)` to push detection results.
 * `detected` and `labels` are auto-derived from the detection list.
 */
export class ObjectSensor<TStorage extends object = Record<string, any>> extends Sensor<ObjectSensorProperties, TStorage> {
  readonly type = SensorType.Object;
  readonly category = SensorCategory.Sensor;

  _requiresFrames = false;

  constructor(name = 'Object Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({
      [ObjectProperty.Detected]: false,
      [ObjectProperty.Detections]: [],
      [ObjectProperty.Labels]: [],
    });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  get detections(): TrackedDetection[] {
    return this.props.detections;
  }

  get labels(): DetectionLabel[] {
    return this.props.labels;
  }

  /**
   * Report detected objects. Auto-derives `detected` and `labels` from the list.
   *
   * - `reportDetections(true)`: something detected without specific data. The SDK
   *   synthesizes a single full-frame `'motion'` detection as a generic fallback.
   * - `reportDetections(true, [...])`: explicit detections (typical case).
   * - `reportDetections(false)`: clear.
   *
   * @param detected - Whether any object is currently detected.
   *
   * @param detections - Optional explicit object detections (with optional tracking metadata).
   *
   * @example
   * ```ts
   * import type { TrackedDetection } from '@camera.ui/sdk';
   * sensor.reportDetections(true, [
   *   {
   *     label: 'person',
   *     confidence: 0.92,
   *     box: { x: 0.1, y: 0.2, width: 0.3, height: 0.4 },
   *   } satisfies TrackedDetection,
   * ]);
   * sensor.reportDetections(false);
   * ```
   */
  reportDetections(detected: boolean, detections?: TrackedDetection[]): void {
    const list = this._normalizeReportedDetections<TrackedDetection>(detected, detections, 'motion');
    const labels = Array.from(new Set(list.map((d) => d.label)));
    this._writeState({
      [ObjectProperty.Detected]: detected,
      [ObjectProperty.Detections]: list,
      [ObjectProperty.Labels]: labels,
    });
  }

  /**
   * Explicitly clear detection state (detected = false, detections = [], labels = []).
   *
   * @example
   * ```ts
   * sensor.clearDetections();
   * ```
   */
  clearDetections(): void {
    this.reportDetections(false);
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Return type for {@link ObjectDetectorSensor.detectObjects}. */
export interface ObjectResult {
  /** Whether any object is detected in this frame. */
  detected: boolean;
  /** Detections emitted for this frame. */
  detections: TrackedDetection[];
}

/**
 * Object detector that receives video frames from the backend pipeline.
 * Extend this class and implement {@link detectObjects} and {@link modelSpec}.
 * The backend scales frames to match `modelSpec.input` dimensions before
 * each call.
 */
export abstract class ObjectDetectorSensor<TStorage extends object = Record<string, any>> extends ObjectSensor<TStorage> {
  override _requiresFrames = true;

  abstract get modelSpec(): ObjectModelSpec;

  /** Analyze a single video frame for objects. Called by the backend at the configured interval. */
  abstract detectObjects(frame: VideoFrameData): Promise<ObjectResult>;
}

/** Registry metadata for {@link ObjectSensor}. */
export const objectMeta = defineSensor({
  type: SensorType.Object,
  category: SensorCategory.Sensor,
  assignmentKey: 'object',
  multiProvider: false,
  isDetectionType: true,
  properties: {
    [ObjectProperty.Detected]: { type: 'boolean' },
    [ObjectProperty.Detections]: { type: 'object', internal: true },
    [ObjectProperty.Labels]: { type: 'object' },
  },
  semantics: null,
});
